common/graphics/texture.h

```diff
@@ -95,7 +95,15 @@ class TextureRegistry {
 
  private:
   std::map<int64_t, std::shared_ptr<Texture>> mapping_;
-  std::map<uintptr_t, std::weak_ptr<ContextListener>> images_;
+  std::atomic_size_t image_counter_;
+  // This map keeps track of registered context listeners by their own
+  // externally provided id. It indexes into ordered_images_.
+  std::map<uintptr_t, size_t> image_indices_;
+  // This map makes sure that iteration of images happens in insertion order
+  // (managed by image_counter_) so that images which depend on other images get
+  // re-created in the right order.
+  std::map<size_t, std::pair<uintptr_t, std::weak_ptr<ContextListener>>>
```

Comment:
I guess this makes sense but is a bit rough to read. How about we take advantage of the fact that the map is ordered already and store the existing key and the order as a key type. Then you can have a custom std::less on the key type and typedef the whole thing for more readibility. Iterating over the map elements will be order then. I suppose you will need a separate map for the removal (a unintptr to order association).

Just a suggestion.

Reply:
I considered that but it comes down to either storing extra data in the key or extra data in the value.

This makes the key easier to hash and avoids the need for a custom comparer that ignores part of the data. I'll use the typedef though.

Reply:
(if I didn't need to worry about updating the order when someone re-registers I could just use a std::vector)